File: src/observability/metrics.ts
export type MetricsSnapshot = Record<string, number>;

function metricName(name: string): string {
  return name.replace(/[^a-zA-Z0-9_]/g, "_");
}

export class MetricsRegistry {
  private readonly counters = new Map<string, number>();

  public increment(name: string, value = 1): void {
    if (!Number.isFinite(value) || value < 0) throw new Error("Incremento de métrica inválido.");
    const key = metricName(name);
    this.counters.set(key, (this.counters.get(key) ?? 0) + value);
  }

  public snapshot(): MetricsSnapshot {
    return Object.fromEntries(this.counters.entries());
  }

  public toPrometheus(): string {
    return [...this.counters.entries()]
      .sort(([left], [right]) => left.localeCompare(right))
      .map(([name, value]) => `${name} ${value}`)
      .join("\n");
  }
}

File: data_platform/python/api/metrics.py
from __future__ import annotations

from threading import Lock


class MetricsRegistry:
    def __init__(self) -> None:
        self._counters: dict[str, float] = {}
        self._lock = Lock()

    def increment(self, name: str, value: float = 1) -> None:
        if value < 0:
            raise ValueError("incremento de métrica inválido")
        metric_name = "".join(character if character.isalnum() or character == "_" else "_" for character in name)
        with self._lock:
            self._counters[metric_name] = self._counters.get(metric_name, 0) + value

    def snapshot(self) -> dict[str, float]:
        with self._lock:
            return dict(self._counters)

    def to_prometheus(self) -> str:
        return "\n".join(f"{name} {value:g}" for name, value in sorted(self.snapshot().items()))
